nep-sales.mediawiki

```diff
@@ -0,0 +1,123 @@
+
+This process is currently not supported by the ecosystem, resulting in several weaknesses in the process:
+
+1. The investor verification process is largely opaque and done off-chain. The source of truth to check if an investor is verified is usually the website of the organsiation. However, it is difficult for the investor to verify that the website and the smart contract is in sync.
+
+2. The investor depends on an announcement for start and closing of the sales from the organisation. However, many people still do not do their research and proceed to participate in a closed sale, causing lost funds and anger.
+
+3. There might be contribution limits that are recorded on the blockchain but it is opaque to the wallet software. Thus, the wallet is unable to stop the investor from over-contributing.
+
+4. There is currently no records of the entire sales process. Tokens can be created at different rates at different times. It is confusing to tell what is the current exchange rate without attempting a transaction first.
+
+This proposal tackles the mentioned issues by proposing a standard set of methods that wallets can implement to provide investors with the visibility and confidence when contributing. With the proposed set of methods, the wallet can independently verify the sale and verification status, providing the investor with confidence in the sales.
+
+It is noted that the above issues are all tackled by using verification and having the contract reject transactions. However, this is a very unfriendly way of rejection with no clarity of the reason of rejection given to the investor. With this proposal, we hope that the token sales experience will be smooth and painless. This proposal is not an excuse to not implement verification as it is still the better option to implement.
+
+==Specification==
+
+We propose the following methods to be implemented in any crowdsale contract that requires investor verification as well as sales restricted to a timeframe.
+
+The methods starting with "kyc" are methods for individuals, indicating their status in the current sale.
+
+The methods starting with "sale" are methods for the entire sale, indicating the status of the current sale.
+
+This standard defines two method types:
+
+* '''(Required)''' : methods that are compulsory in order to be NEP-compliant. Methods are required unless marked otherwise. These methods are mostly user-facing and thus must be standardised for a smooth experience.
+
+* '''(Optional)''' : methods that are optionally implemented. These methods are not all user-facing and contains certains parameters not present in all sales. However, developer tools are encouraged to follow this specification for the best experience.
+
+===Methods===
+
+====mintTokens====
+
+* Syntax: <code>public static string mintTokens(byte[] address)</code>

```

Comment:
What is the `string` return type for? It would be cool if it was possible to return a `string` message when a mint fails, but I'm not sure that's feasible given how raw transactions are submitted, and also in the context of `TriggerType.Verification`. It's worth considering if there is a way to return a human-readable message, and if not, changing this to just return `bool`.

Reply:
I'm not well versed in this area so I need to ask if the return value appears anywhere else other than through RPC `invoke`? I think that return values are a little wasted currently... I do like if we can return something more relevant than just a `bool`.